### DISCUSSION TITLE
关于可以使用的第三方库

### DISCUSSION THREAD

**Discussion by liusy58** [liusy58/NJU_AdvancedProgramming_SP22#66] [Announcements]
每一位同学需要使用的第三方库都需要向助教请示，并给出具体原因，基础功能原则上不允许使用第三方库，扩展功能可以。

**Comment by binbin00** [comment]
第三方库：
https://chariri.moe/archives/408
https://chariri.moe/archives/392
https://github.com/cqjjjzr/utf8-console
https://github.com/liusy58/NJU_AdvancedProgramming_SP22

原因：
本项目需要处理汉字，那就需要考虑编码问题。正常情况下应该整个程序 UTF-8 化，但由于该禁令，至少人为引入了三个（不必要的）难点：
1. 在控制台下中文文本的对齐。最朴素的 std::setw 当然是不能直接用的，中文可能用 2 个字符，可能用 3 个字符。光是实现了简单的一个解决方案就用了两组第三方的代码（UTF8-CPP 和一个 portable 的 wcwidth 实现），且我已经将该方案 Pull Request 到本 repo。如果因此我要被判定使用了两个库，其他同学用我的代码被判定使用了三个库，无疑是相当滑稽的。
2. 判定中文字符以计算长度。Discussion 中提到了：我们不能用 length 函数计算长度，应该判定真实长度，即我们就需要判断一段字符串中 code point 的数量而不是 code unit 的数量。以 UTF-8 为例，这必须要使用 UTF8-CPP 、icu 或 iconv 等处理库。否则就需要自己写字符边界检测（见下面的理由 2）——极为脏的做法。我们不应该让代码里充斥着这种和业务无关的轮子；
3. 在 Windows 系统下，如果系统的 Code Page 没有切换到 UTF-8 即 65001（冷知识：Windows 直到 Windows 10 1803 才支持把全局 Code Page 改为 UTF-8），需要至少使用：a) system("chcp 65001"); b) SetConsoleCP 与 SetConsoleOutputCP 变更控制台 CodePage。后者就要用到 Windows.h——被禁止使用了！
